### PR TITLE
Update regex for various date times per CCB-350

### LINKED ISSUE #520: Refactor regex handling to use Google's RE2J library and compile
## 💪 Motivation
Per https://github.com/NASA-PDS/validate/security/code-scanning/8, our software could be significantly more efficient using the Google library instead of standard Java libraries.

Additionally, per comments from @al-niessner, we should be compiling the regex prior to checking for a match to improve performance. Not sure if that is still the case if we switch to Google's regex library.

## ⚖️ Acceptance Criteria
The short answer: regression tests all pass

Some more specific criteria examples:

**Given** some valid PDS4 bundle
**When I perform** validation
**Then I expect** the bundle to pass as expected, ensuring it crawls through and validating all files as expected (the regex is used to find all XML files in the bundle)

**Given** and invalid PDS4 bundle with at least 1 invalid LID/LIDVID
**When I perform** validation
**Then I expect** that bundle to fail indicating an invalid LID/LIDVID

### LINKED ISSUE #607: Support for Improved Datetime Checks
## 💡 Description

Per [CCB-350](https://pds-jira.jpl.nasa.gov/browse/CCB-350)

### LINKED ISSUE #608: Update datetime regex for content validation
## 💡 Description

New regex needed for datetime validation for table data. Here are the updated schematron patterns that will need to be translated into Java regex:

- [ ] ASCII_Date_DOY -- updates
```
    <xs:simpleType name="ASCII_Date_DOY">
      <xs:restriction base="xs:string">
        <xs:pattern value='(-)?[0-9]{4}(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(Z?)'/>
      </xs:restriction>
    </xs:simpleType>
```

- [ ] ASCII_Date_Time_DOY -- updates
```
    <xs:simpleType name="ASCII_Date_Time_DOY">
      <xs:restriction base="xs:string">
        <xs:pattern value='(-)?[0-9]{4}(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(1981|1982|1983|1985|1993|1994|1997|2015)-181T23:59:60(\.[0-9]+)?(Z?)'/>
        <xs:pattern value='(1972|1992|2012)-182T23:59:60(\.[0-9]+)?(Z?)'/>
        <xs:pattern value='(1971|1973|1974|1975|1977|1978|1979|1987|1989|1990|1995|1998|2005)-365T23:59:60(\.[0-9]+)?(Z?)'/>
        <xs:pattern value='(1972|1976|2008|2016)-366T23:59:60(\.[0-9]+)?(Z?)'/>
      </xs:restriction>
    </xs:simpleType>
```

- [ ] ASCII_Date_Time_DOY_UTC -- updates	
```
   <xs:simpleType name="ASCII_Date_Time_DOY_UTC">   
      <xs:restriction base="xs:string">
        <xs:pattern value='(-)?[0-9]{4}(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((00[1-9])|(0[1-9][0-9])|([1-2][0-9][0-9])|(3(([0-5][0-9])|(6[0-5]))))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(1981|1982|1983|1985|1993|1994|1997|2015)-181T23:59:60(\.[0-9]+)?(Z)'/>
        <xs:pattern value='(1972|1992|2012)-182T23:59:60(\.[0-9]+)?(Z)'/>
        <xs:pattern value='(1971|1973|1974|1975|1977|1978|1979|1987|1989|1990|1995|1998|2005)-365T23:59:60(\.[0-9]+)?(Z)'/>
        <xs:pattern value='(1972|1976|2008|2016)-366T23:59:60(\.[0-9]+)?(Z)'/>
      </xs:restriction>
    </xs:simpleType>
```

- [ ] ASCII_Date_Time_YMD -- updates
```
   <xs:simpleType name="ASCII_Date_Time_YMD">
      <xs:restriction base="xs:string">
        <xs:pattern value='(-)?[0-9]{4}(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((0[1-9])|(1[0-2]))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(T)(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='((1971|1972|1973|1974|1975|1976|1977|1978|1979|1987|1989|1990|1995|1998|2005|2008|2016)-12-31T23:59:60(\.[0-9]+)?(Z?)|(1972|1981|1982|1983|1985|1992|1993|1994|1997|2012|2015)-06-30T23:59:60(\.[0-9]+)?(Z?))'/>
      </xs:restriction>
    </xs:simpleType>
```

- [ ] ASCII_Date_Time_YMD_UTC -- updates	
```
    <xs:simpleType name="ASCII_Date_Time_YMD_UTC">
      <xs:restriction base="xs:string">
        <xs:pattern value='(-)?[0-9]{4}(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((0[1-9])|(1[0-2]))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(T)(([0-1][0-9])|(2[0-3]))(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31)))(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z)'/>
        <xs:pattern value='((1971|1972|1973|1974|1975|1976|1977|1978|1979|1987|1989|1990|1995|1998|2005|2008|2016)-12-31T23:59:60(\.[0-9]+)?(Z)|(1972|1981|1982|1983|1985|1992|1993|1994|1997|2012|2015)-06-30T23:59:60(\.[0-9]+)?(Z))'/>
      </xs:restriction>
    </xs:simpleType>
```

- [ ] ASCII_Date_YMD -- updates	
```
    <xs:simpleType name="ASCII_Date_YMD">
      <xs:restriction base="xs:string">
        <xs:pattern value='(-)?[0-9]{4}(Z?)'/>
        <xs:pattern value='(-)?[0-9]{4}-((0[1-9])|(1[0-2]))(Z?)'/>
        <xs:pattern value='((-)?[0-9]{4}-(02-((0[1-9])|(1[0-9])|(2[0-8])))|(-)?[0-9]{4}-((04|06|09|11)-((0[1-9])|([1-2][0-9])|30))|(-)?[0-9]{4}-((01|03|05|07|08|10|12)-((0[1-9])|([1-2][0-9])|(30|31))))(Z?)'/>
        <xs:pattern value='(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)-02-29)(Z?)'/>
        <xs:pattern value='(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(Z?)'/>
      </xs:restriction>
    </xs:simpleType>
```

- [ ] ASCII_Time -- updates		
```
    <xs:simpleType name="ASCII_Time">
      <xs:restriction base="xs:string">
        <xs:pattern value='(([0-1][0-9])|(2[0-3]))(Z?)'/>
        <xs:pattern value='(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)'/>
        <xs:pattern value='(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\.([0-9]+))?(Z?)'/>
        <xs:pattern value='(23):[0-5][0-9]:(([0-5][0-9])|60)((\.[0-9]+))?(Z?)'/>
      </xs:restriction>
    </xs:simpleType>
```

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Put XML in Java then did some find-replace.

## ⚙️ Test Data and/or Report
See unit tests

## ♻️ Related Issues
Closes #608 
Closes #607 
Closes #520

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DateTimeValidator.java [hunk 1]
```diff
@@ -31,76 +31,138 @@ public class DateTimeValidator {
...
+      Pattern.compile("(-)?([0-9]{2}(04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\\.([0-9]+))?(Z)"),
+      Pattern.compile("(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(Z)"),
+      Pattern.compile("(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3]))(Z)"),
+      Pattern.compile("(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z)"),
+      Pattern.compile("(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-366(T)(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\\.([0-9]+))?(Z)"),
+      Pattern.compile("(1981|1982|1983|1985|1993|1994|1997|2015)-181T23:59:60(\\.[0-9]+)?(Z)"),
+      Pattern.compile("(1972|1992|2012)-182T23:59:60(\\.[0-9]+)?(Z)"),
+      Pattern.compile("(1971|1973|1974|1975|1977|1978|1979|1987|1989|1990|1995|1998|2005)-365T23:59:60(\\.[0-9]+)?(Z)"),
+      Pattern.compile("(1972|1976|2008|2016)-366T23:59:60(\\.[0-9]+)?(Z)"));
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DateTimeValidator.java:84]: @sonatype-lift ignore

**Comment by al-niessner** [comment]
@jordanpadams @nutjob4life @tloubrieu-jpl 

Ready for review. I left it with the Java standard regex since it is 4x faster than google RE2J when compiled first.

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DateTimeValidator.java [hunk 2]
```diff
@@ -31,76 +31,138 @@ public class DateTimeValidator {
...
+      Pattern.compile("(-)?(00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96)00-02-29(Z?)"));
 
-  private static List<String> TIME_FORMATS = Arrays.asList("(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)",
-      "(([0-1][0-9])|(2[0-3])):[0-5][0-9]:(([0-5][0-9])|60)((\\.[0-9]+)|)(Z?)",
-      "(([0-1][0-9])|(2[0-4]))(Z?)", "24:00((:00((\\.0+)|))|)(Z?)");
+  private static List<Pattern> TIME_FORMATS = Arrays.asList(
+      Pattern.compile("(([0-1][0-9])|(2[0-3]))(Z?)"),
+      Pattern.compile("(([0-1][0-9])|(2[0-3])):[0-5][0-9](Z?)"),
+      Pattern.compile("(([0-1][0-9])|(2[0-3])):[0-5][0-9]:([0-5][0-9])(\\.([0-9]+))?(Z?)"),
+      Pattern.compile("(23):[0-5][0-9]:(([0-5][0-9])|60)((\\.[0-9]+))?(Z?)"));
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/rule/pds4/DateTimeValidator.java:159]: You even got leap seconds in there! That's some crazy close attention to detail!

File: src/test/java/gov/nasa/pds/validate/DateTimeEfficiency.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  public static void main(String[] args) {
+    try {
+      long t0 = System.currentTimeMillis();
+      for (int i=0 ; i < 1000000 ; i++) {
+        DateTimeValidator.isValid(FieldType.ASCII_DATE_TIME_DOY_UTC, "2023-120T10:23:55.123456Z");
+      }
+      System.out.println("Processing time: " + (System.currentTimeMillis() - t0)/1000. + " seconds");
+    } catch (Exception e) {
+      System.err.println("Complete and utter failure.");
```
Comment (nutjob4life) [inline src/test/java/gov/nasa/pds/validate/DateTimeEfficiency.java:16]: 😅

**Review by nutjob4life — APPROVED** [review]
LGTM 👍

**Pull request merged**